File: problems/Valid Parenthesis String/checkValidString.cs
public class Solution {
    public bool CheckValidString(string s) {
        var l = 0;
        var r = 0;

        foreach (var letter in s) {
            l += '(' == letter ? 1 : -1;
            r += ')' != letter ? 1 : -1;

            if (0 > r) {
                break;
            }

            l = Math.Max(0, l);
        }

        return 0 == l;
    }
}

File: problems/Valid Parenthesis String/checkValidString.java
class Solution {
    public boolean checkValidString(String s) {
        int openParenthesis = 0;
        int closeParenthesis = 0;

        for (char letter : s.toCharArray()) {
            openParenthesis += '(' == letter ? 1 : -1;
            closeParenthesis += ')' != letter ? 1 : -1;

            if (0 > closeParenthesis) {
                break;
            }

            openParenthesis = Math.max(0, openParenthesis);
        }

        return 0 == openParenthesis;
    }
}
